Crash while running WebCore::SourceBufferPrivate::processMediaSamplesOperation

jyavenard opened the pull request:
#### c54dbe8ebd21d9268a73bb4e611bf9424d8172e7
<pre>
Crash while running WebCore::SourceBufferPrivate::processMediaSamplesOperation
<a href="https://bugs.webkit.org/show_bug.cgi?id=265008">https://bugs.webkit.org/show_bug.cgi?id=265008</a>
<a href="https://rdar.apple.com/116953621">rdar://116953621</a>

Reviewed by Youenn Fablet.

It was possible for either memoryPressure or removeCodedFrame to be run in the middle of a pending appendBuffer operation.
The move to NativePromise already eliminated most of the possibilities by removeCodedFrame operation wasn&apos;t queued to the pending
list of operation.
We now only process removeCodedFrame once we have fully completed any pending appendBuffer.
Also, we allow for the appendBuffer operation to be aborted in more places.

* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::removeCodedFrames):
(WebCore::SourceBufferPrivate::didReceiveInitializationSegment):
(WebCore::SourceBufferPrivate::append):
(WebCore::SourceBufferPrivate::processPendingMediaSamples):

Canonical link: <a href="https://commits.webkit.org/270929@main">https://commits.webkit.org/270929@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a896fb6e660c4e30cd052a0884ab738d328f42bc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26816 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5432 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28052 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29027 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24510 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27262 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7266 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2823 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24400 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27078 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4245 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23014 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3724 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3771 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24011 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29507 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24455 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24412 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30013 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3804 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1999 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27921 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5257 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6448 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4271 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4166 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -408,6 +408,22 @@ MediaTime SourceBufferPrivate::findPreviousSyncSamplePresentationTime(const Medi
 Ref<MediaPromise> SourceBufferPrivate::removeCodedFrames(const MediaTime& start, const MediaTime& end, const MediaTime& currentTime)
+{
+    MediaPromise::Producer producer;
+    Ref<MediaPromise> promise = producer.promise();
```
`Ref` suffices.

youennf commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -408,6 +408,22 @@ MediaTime SourceBufferPrivate::findPreviousSyncSamplePresentationTime(const Medi
+        producer.settle(result);
+        return MediaPromise::createAndSettle(WTFMove(result));
+    });
+    return promise;
```
Can we return m_currentSourceBufferOperation instead, it seems equivalent to a new promise that gets settled when m_currentSourceBufferOperation gets settled.

jyavenard replied:
I would need to make MediaPromise non exclusive. 

I hesitated in doing so, but knowing your reluctance on even having a non exclusive mode. I decided against it. 

With a non exclusive promise we could indeed make the code simpler.

jyavenard replied:
I thought of a solution, which I think is quite elegant https://bugs.webkit.org/show_bug.cgi?id=265096

youennf commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -581,27 +597,32 @@ void SourceBufferPrivate::didReceiveInitializationSegment(InitializationSegment&
 
-    m_currentAppendProcessing = m_currentAppendProcessing->whenSettled(RunLoop::current(), [segment = WTFMove(segment), weakThis = WeakPtr { *this }, this](auto&& result) mutable -> Ref<MediaPromise> {
+    auto segmentCopy = segment;
+    m_currentAppendProcessing = m_currentAppendProcessing->whenSettled(RunLoop::current(), [segment = WTFMove(segment), weakThis = WeakPtr { *this }, this, abortCount = m_abortCount](auto&& result) mutable {
```
Do we need segmentCopy if we just copy segment instead of moving it in this lambda?

jyavenard replied:
It was a matter of having just 1 copy vs 2.

jyavenard replied:
oh I see what you mean.

The issue is in the order of running of the capture. 
the segment needs to be copied and then moved later.
but when doing 
`blah([obj = WTFMove(obj] { })->blah2([obj = WTFMove(obj) { } );`

where is the move happening first?
C++17 defines that when you do F(A), F is evaluated before A. But at a glance, I find it super dodgy. 
so in theory:
`blah([obj] { })->blah2([obj = WTFMove(obj) { } );` should do what we want.
But I find it much clearer to explicitly take the copy first and move that copy.

jyavenard committed:
Crash while running WebCore::SourceBufferPrivate::processMediaSamplesOperation
https://bugs.webkit.org/show_bug.cgi?id=265008
rdar://116953621

Reviewed by Youenn Fablet.

It was possible for either memoryPressure or removeCodedFrame to be run in the middle of a pending appendBuffer operation.
The move to NativePromise already eliminated most of the possibilities by removeCodedFrame operation wasn't queued to the pending
list of operation.
We now only process removeCodedFrame once we have fully completed any pending appendBuffer.
Also, we allow for the appendBuffer operation to be aborted in more places.

* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::removeCodedFrames):
(WebCore::SourceBufferPrivate::didReceiveInitializationSegment):
(WebCore::SourceBufferPrivate::append):
(WebCore::SourceBufferPrivate::processPendingMediaSamples):

Canonical link: https://commits.webkit.org/270929@main
